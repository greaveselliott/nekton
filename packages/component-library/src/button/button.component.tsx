import React from 'react';
import classNames from 'classnames';
import './button.scss';

export interface IProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement | HTMLAnchorElement> {
  shape?: 'default' | 'round';
  variant?: 'primary' | 'secondary' | 'negative';
  size?: 'large' | 'medium' | 'small';
  children?: React.ReactChild | React.ReactChild[];
  disabled?: boolean;
  href?: string;
  onClick?: () => void;
}

const Button: React.FC<IProps> = ({
  children,
  variant = 'primary',
  shape = 'default',
  size,
  href = '',
  ...props
}) => {
  const Element = href ? 'a' : 'button';

  return (
    <Element
      {...props}
      className={classNames({
        button: true,
        [`button--${variant}`]: variant,
        [`button--${shape}`]: shape,
        [`button--round--${size}`]: size
      })}
    >
      {children}
    </Element>
  );
};

export default Button;
