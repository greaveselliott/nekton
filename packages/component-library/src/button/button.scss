@import '../global';

$button-font: $global-font-family-primary;
$button-disabled-background-color: $global-color-disabled;
$button-disabled-text-color: $global-color-shade-three;
$button-disabled-border-color: $global-color-disabled;
$button-primary-text-color: $global-color-shade-one;
$button-primary-background-color: $global-color-primary;
$button-primary-border-color: $global-color-primary;
$button-primary-hover-background-color: $global-color-primary-dark;
$button-secondary-text-color: $global-color-primary;
$button-secondary-hover-background-color: $global-color-primary-muted;
$button-secondary-hover-border-color: $global-color-primary-muted;
$button-secondary-hover-text-color: $global-color-shade-two;
$button-negative-border-color: $global-color-negative;
$button-negative-text-color: $global-color-negative;
$button-negative-hover-border-color: $global-color-negative;
$button-negative-hover-background-color: $global-color-negative-muted;

.button {
  border-style: solid;
  border-width: 1px;
  font-weight: bold;
  font-size: 14px;
  line-height: 18px;
  box-sizing: border-box;
  font-family: $button-font;
  text-decoration: none;
  cursor: pointer;

  &:disabled {
    background-color: $button-disabled-background-color;
    color: $button-disabled-text-color;
    border-color: $button-disabled-border-color;
    cursor: not-allowed;
  }
  &--primary {
    color: $button-primary-text-color;
    background-color: $button-primary-background-color;
    border-color: $button-primary-text-color;
    &:hover:not([disabled]) {
      background-color: $button-primary-hover-background-color;
      color: $button-primary-text-color;
    }
  }

  &--secondary {
    border-color: $button-primary-border-color;
    color: $button-secondary-text-color;
    &:hover:not([disabled]) {
      background-color: $button-secondary-hover-background-color;
      color: $button-secondary-hover-text-color;
      border-color: $button-secondary-hover-border-color;
    }
  }

  &--default {
    padding: 10px 14px;
    border-radius: 50px;
    width: 100%;
    height: 40px;
  }

  &--negative {
    border-color: $button-negative-border-color;
    color: $button-negative-text-color;

    &:hover:not([disabled]) {
      background-color: $button-negative-hover-background-color;
      border-color: $button-negative-hover-border-color;
    }
  }

  &--round {
    border-radius: 50%;

    &--small {
      height: 30px;
      width: 30px;
    }

    &--medium {
      height: 52px;
      width: 52px;
    }

    &--large {
      height: 70px;
      width: 70px;
    }
  }
}
