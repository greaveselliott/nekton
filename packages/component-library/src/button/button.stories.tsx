import React from 'react';
import Button from './button.tsx';

export default { title: 'Button' };

export const defaultPrimaryButtonWithText = () => (
  <Button style="primary">
    <span>Primary Button</span>
  </Button>
);

export const defaultPrimaryDisabledButtonWithText = () => (
  <Button style="primary" disabled={true}>
    <span>Primary Button</span>
  </Button>
);

export const defaultSecondaryButtonWithText = () => (
  <Button type="default" style="secondary">
    <span>Secondary Button</span>
  </Button>
);

export const smallRoundPrimaryButton = () => (
  <Button type="round" style="primary" size="small"></Button>
);

export const smallRoundSecondaryButton = () => (
  <Button type="round" style="secondary" size="small"></Button>
);

export const smallRoundDisabledButton = () => (
  <Button type="round" style="secondary" size="small" disabled={true}></Button>
);

export const mediumRoundPrimaryButton = () => (
  <Button type="round" style="primary" size="medium"></Button>
);

export const mediumRoundSecondaryButton = () => (
  <Button type="round" style="secondary" size="medium"></Button>
);

export const mediumRoundSecondaryDisabledButton = () => (
  <Button type="round" style="secondary" size="medium" disabled={true}></Button>
);

export const largeRoundPrimaryButton = () => (
  <Button type="round" style="primary" size="large"></Button>
);

export const largeRoundSecondaryButton = () => (
  <Button type="round" style="secondary" size="large"></Button>
);

export const smallRoundNegativeButton = () => (
  <Button type="round" style="negative" size="small"></Button>
);

export const smallRoundNegativeDisabledButton = () => (
  <Button type="round" style="negative" size="small" disabled={true}></Button>
);

export const mediumRoundNegativeButton = () => (
  <Button type="round" style="negative" size="medium"></Button>
);

export const mediumRoundNegativeDisabledButton = () => (
  <Button type="round" style="negative" size="medium" disabled={true}></Button>
);

export const largeRoundNegativeButton = () => (
  <Button type="round" style="negative" size="large"></Button>
);

export const largeRoundNegativeDisabledButton = () => (
  <Button type="round" style="negative" size="large" disabled={true}></Button>
);

export const defaultPrimaryButtonWithLink = () => (
  <Button href="#">Some Text</Button>
);
