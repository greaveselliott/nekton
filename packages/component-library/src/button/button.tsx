import React from 'react';
import classNames from 'classnames';
import './button.scss';

type props = {
  type?: 'default' | 'round';
  style?: 'primary' | 'secondary' | 'negative';
  size?: 'large' | 'medium' | 'small';
  disabled?: boolean;
  children?: React.ReactNode;
  onClick?: () => void;
  href?: string;
};

const Button: React.FC<props> = ({
  children,
  style,
  disabled,
  type,
  size,
  onClick,
  href
}) => {
  const Element = href ? 'a' : 'button';

  return (
    <Element
      className={classNames({
        button: true,
        [`button--${style}`]: style,
        [`button--${type}`]: type,
        [`button--round--${size}`]: size
      })}
      disabled={disabled}
      onClick={onClick}
      href={href}
    >
      {children}
    </Element>
  );
};

export default Button;

Button.defaultProps = {
  type: 'default',
  style: 'primary'
};
