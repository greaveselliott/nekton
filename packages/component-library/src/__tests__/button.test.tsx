import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { Button } from '../button';

describe('Button', () => {
  it('renders as expected', () => {
    const button = mount(
      <Button type="round" style="secondary" size="small" />
    );
    expect(button).toMatchSnapshot();
  });

  it('renders a link as expected', () => {
    const button = mount(<Button href="#" type="default" style="secondary" />);
    expect(button).toMatchSnapshot();
  });

  it('renders a button elemement as default', () => {
    const button = shallow(<Button />);
    expect(button).toContainMatchingElement('button');
  });

  it('renders an anchor tag if given a href', () => {
    const button = shallow(<Button href="#" />);
    expect(button).toContainMatchingElement('a');
  });

  it('is able to call a function on click', () => {
    const mockOnClickFunction = jest.fn();
    const button = shallow(<Button onClick={mockOnClickFunction} />);
    button.simulate('click');
    expect(mockOnClickFunction).toHaveBeenCalled();
  });

  it('can be disabled', () => {
    const button = shallow(<Button disabled={true} />);
    expect(button).toBeDisabled();
  });

  it('can render children', () => {
    const buttonText = 'Search';
    const button = shallow(<Button disabled={true}>{buttonText}</Button>);
    expect(button).toIncludeText(buttonText);
  });
});
