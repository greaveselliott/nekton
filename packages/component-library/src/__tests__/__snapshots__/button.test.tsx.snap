// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a link as expected 1`] = `
<Button
  href="#"
  style="secondary"
  type="default"
>
  <a
    className="button button--secondary button--default"
    href="#"
  />
</Button>
`;

exports[`Button renders as expected 1`] = `
<Button
  size="small"
  style="secondary"
  type="round"
>
  <button
    className="button button--secondary button--round button--round--small"
  />
</Button>
`;
